src/tests/test_server_stream.cpp

```diff
@@ -0,0 +1,520 @@
+using error_code = boost::system::error_code;
+using ssl_stream = Botan::TLS::Stream<net::ip::tcp::socket>;
+using namespace std::placeholders;
+using Result = Botan_Tests::Test::Result;
+
+static const auto k_timeout = std::chrono::seconds(3);
+static const auto k_endpoints = std::vector<tcp::endpoint> {tcp::endpoint{net::ip::make_address("127.0.0.1"), 8082}};
+
+enum { max_msg_length = 512 };
+
+static std::string server_cert() { return Botan_Tests::Test::data_dir() + "/x509/certstor/cert1.crt"; }
+static std::string server_key() { return Botan_Tests::Test::data_dir() + "/x509/certstor/key01.pem"; }
+
+class Timeout_Exception : public std::runtime_error
+   {
+      using std::runtime_error::runtime_error;
+   };
+
+struct Side
+   {
+      Side()
+         : m_credentials_manager(true, ""),
+           m_ctx(m_credentials_manager, m_rng, m_session_mgr, m_policy, Botan::TLS::Server_Information()) {}
+
+      Side(const std::string& server_cert, const std::string& server_key)
+         : m_credentials_manager(m_rng, server_cert, server_key),
+           m_ctx(m_credentials_manager, m_rng, m_session_mgr, m_policy, Botan::TLS::Server_Information()) {}
+
+      virtual ~Side() {}
+
+      net::mutable_buffer buffer() { return net::buffer(m_data, max_msg_length); }
+      net::mutable_buffer buffer(size_t size) { return net::buffer(m_data, size); }
+
+      std::string message() const { return std::string(m_data); }
+
+   protected:
+      Botan::AutoSeeded_RNG m_rng;
+      Basic_Credentials_Manager m_credentials_manager;
+      Botan::TLS::Session_Manager_Noop m_session_mgr;
+      Botan::TLS::Policy m_policy;
+      Botan::TLS::Context m_ctx;
+      std::unique_ptr<ssl_stream> m_stream;
+
+      char m_data[max_msg_length];
+   };
+
+struct Result_Wrapper
+   {
+      Result_Wrapper(net::io_context& ioc, const std::string& name) : m_timer(ioc), m_result(name) {}
+
+      Result& result() { return m_result; }
+
+      void set_timer(const std::string& msg)
+         {
+         m_timer.expires_after(k_timeout);
+         m_timer.async_wait([this, msg](const error_code &ec)
+            {
+            if(ec != net::error::operation_aborted)  // timer cancelled
+               {
+               m_result.test_failure(m_result.who() + ": timeout in " + msg);
+               throw Timeout_Exception(m_result.who());
+               }
+            });
+         }
+
+      void stop_timer()
+         {
+         m_timer.cancel();
+         }
+
+      void check_ec(const std::string& msg, const error_code& ec)
+         {
+         if(ec)
+            { m_result.test_failure(msg, ec.message()); }
+         else
+            { m_result.test_success(msg); }
+         }
+
+      void confirm(const std::string& msg, bool condition)
+         {
+         m_result.confirm(msg, condition);
+         }
+
+   private:
+      net::system_timer m_timer;
+      Result m_result;
+   };
+
+class Server : public Side, public std::enable_shared_from_this<Server>
```

Comment:
The server side is rather passive in all tests. It almost always performs the same checks. `expect_short_read` and `shutdown` allow tests cases to change it's behavior when needed.